lib/nethealth/nethealth_test.go

```diff
@@ -0,0 +1,343 @@
+				"peer-3": {
+					name:             "peer-3",
+					lastStatusChange: server.clock.Now(),
+					addr:             &net.IPAddr{},
+				},
+			},
+		},
+		{
+			description: "remove peer 2 and 3 from the cluster",
+			rm:          []string{"peer-2", "peer-3"},
+			expected: map[string]*peer{
+				"peer-1": {
+					name:             "peer-1",
+					lastStatusChange: server.clock.Now(),
+					addr:             &net.IPAddr{},
+				},
+			},
+		},
+	}
+
+	for _, tt := range cases {
+		for _, node := range tt.add {
+			_, err := server.client.CoreV1().Nodes().Create(node)
+			assert.NoError(t, err, tt.description)
+		}
+
+		for _, node := range tt.rm {
+			err := server.client.CoreV1().Nodes().Delete(node, nil)
+			assert.NoError(t, err, tt.description)
+		}
+
+		err := server.resyncPeerList()
+		assert.NoError(t, err, tt.description)
+		assert.Equal(t, tt.expected, server.peers, tt.description)
+	}
+
+}
+
+func TestResyncPods(t *testing.T) {
+	server := testServer(t)
+	nodes := []*v1.Node{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node-1",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "peer-1",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "peer-2",
+			},
+		},
+	}
+
+	cases := []struct {
+		add            []*v1.Pod
+		rm             []string
+		expectedPeers  map[string]*peer
+		expectedLookup map[string]string
+		description    string
+	}{
+		{
+			description: "ignore our own node",
+			add: []*v1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "pod-node-1",
+						Labels: map[string]string{
+							"k8s-app": "nethealth",
+						},
+					},
+					Spec: v1.PodSpec{
+						NodeName: "node-1",
+					},
+					Status: v1.PodStatus{
+						PodIP: "1.1.1.1",
+					},
+				},
+			},
+			expectedPeers: map[string]*peer{
+				"peer-1": {
+					name:             "peer-1",
+					lastStatusChange: server.clock.Now(),
+					addr:             &net.IPAddr{},
+				},
+				"peer-2": {
+					name:             "peer-2",
+					lastStatusChange: server.clock.Now(),
+					addr:             &net.IPAddr{},
+				},
+			},
+			expectedLookup: map[string]string{},
+		},
+		{
+			description: "new pod for peer-1",
+			add: []*v1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "pod-peer-1",
+						Labels: map[string]string{
+							"k8s-app": "nethealth",
+						},
+					},
+					Spec: v1.PodSpec{
+						NodeName: "peer-1",
+					},
+					Status: v1.PodStatus{
+						PodIP: "1.1.1.1",
+					},
+				},
+			},
+			expectedPeers: map[string]*peer{
+				"peer-1": {
+					name:             "peer-1",
+					lastStatusChange: server.clock.Now(),
+					addr: &net.IPAddr{
+						IP: net.ParseIP("1.1.1.1"),
+					},
+				},
+				"peer-2": {
+					name:             "peer-2",
+					lastStatusChange: server.clock.Now(),
+					addr:             &net.IPAddr{},
+				},
+			},
+			expectedLookup: map[string]string{
+				"1.1.1.1": "peer-1",
+			},
+		},
+		{
+			description: "pod that doesn't have a node set",
```

Comment:
Can you elaborate what exactly is expected in this test case - I don't see a difference in `expectedLookup` and would like to know where to look.

Reply:
When a pod is first created, it may not have the NodeName set on the pod object. I don't expect this should occur with a daemonset, but usually it's worth validating this behavior since it's an external system. So this test simply ensures that if the kubernetes object isn't in an expected state, that it doesn't cause unexpected results in the internal structure.